discv5/src/discv5.nim

```diff
@@ -0,0 +1,85 @@
+import
+  random, chronos, sequtils, chronicles, tables, stint, options, std/bitops,
+  eth/[keys, rlp, async_utils], eth/p2p/enode, eth/trie/db,
+  eth/p2p/discoveryv5/[discovery_db, enr, node, types, routing_table, encoding],
+  eth/p2p/discoveryv5/protocol as discv5_protocol,
+  ./utils
+
+const
+    N = 100
+    MAX_LOOKUPS = 10
+    RUNS = 10
+
+proc randNode(nodes: seq[discv5_protocol.Protocol]): Node =
+    randomize()
+    result = nodes[rand(N - 1)].localNode
+
+proc randNodeFromNodes(nodes: seq[Node]): Node =
+    randomize()
+    result = nodes[rand(nodes.len - 1)]
+
+proc runWith(node: discv5_protocol.Protocol, nodes: seq[discv5_protocol.Protocol]) {.async.} =
+    let target = randNode(nodes)
+    let tid = recordToNodeID(target.record)
+
+    var peer: Node
+    while true:
+        peer = randNode(nodes)
+        if peer.record.toUri() != target.record.toUri():
+            break
+
+    var distance = distanceTo(recordToNodeID(peer.record), tid)
+
+    var called = newSeq[string](0)
+
+    for i in 0..<MAX_LOOKUPS:
+        let lookup = await node.findNode(peer, distance)
+        called.add(peer.record.toUri())
+
+        if lookup.len == 0:
+            distance = 256
+            continue
+
+        for n in items(lookup):
+            let uri = n.record.toUri()
+            if uri == target.record.toUri():
+                echo "Found target in ", i + 1, " lookups"
+                return
+
+            if containsNodeId(called, uri):
+                continue
+
+            let d = distanceTo(recordToNodeID(n.record), tid)
+            if d < distance:
+                peer = n
+                distance = d
+
+        while true: # This ensures we get a random node from the last lookup if we have already called the new peer.
+            if not containsNodeId(called, peer.record.toUri()):
+                break
+
+            peer = randNodeFromNodes(lookup)
+
+    echo "Not found in max iterations"
+
+proc run() {.async.} =
+    var nodes = newSeq[discv5_protocol.Protocol](0)
+
+    for i in 0..<N:
+        let node = initDiscoveryNode(newPrivateKey(), localAddress(20300 + i), if i > 0: @[nodes[0].localNode.record] else: @[])
+        node.start()
```

Comment:
How do these N nodes discover each other initially?

Reply:
The first node acts as bootstrap for the N-1 others.
The `node.start()` line will start the lookup loop and revalidation loop. The first should run a self lookup and a random lookup every minute. However, I'm not so sure that with the 50 second sleep of below, every node of the N nodes actually manages to complete that. And even if it does, 1 run is perhaps a little low. So the DHT of some nodes might not be populated enough yet.

Reply:
@kdeme I also tried with 5 minutes, it didn't make much difference to the result.

Reply:
Are you logging those nodes?